java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java

```diff
@@ -0,0 +1,298 @@
+                        startPos = (int)sp.getStartPosition(info.getCompilationUnit(), tree);
+                    }
+                    String headerText = info.getText().substring(startPos, startOffset);
+                    int idx = headerText.indexOf('{'); //NOI18N
+                    if (idx < 0) {
+                        treeKindCtx = null;
+                        stringCtx = CLASS_HEADER;
+                    }
+                    break;
+                case FOR_LOOP:
+                case ENHANCED_FOR_LOOP:
+                    if (!isRightParenthesisOfLoopPresent(info, startOffset)) {
+                        treeKindCtx = null;
+                    }
+                    break;
+                case PARENTHESIZED:
+                    if (isPartOfWhileLoop(info, startOffset)) {
+                        if (!isRightParenthesisOfLoopPresent(info, startOffset)) {
+                            treeKindCtx = null;
+                        }
+                    }
+                    break;
+            }
+        }
+        List<CodeAction> codeActions = new ArrayList<>();
+        for (CodeTemplate codeTemplate : CodeTemplateManager.get(doc).getCodeTemplates()) {
+            String parametrizedText = codeTemplate.getParametrizedText();
+            if (parametrizedText.toLowerCase().contains(SELECTION_VAR) && !TO_FILTER.contains(codeTemplate.getAbbreviation())) {
+                if (codeTemplate.getContexts() == null || codeTemplate.getContexts().isEmpty() || accept(codeTemplate, treeKindCtx, stringCtx)) {
+                    String label = html2text(codeTemplate.getDescription());
+                    if (label == null) {
+                        String text = codeTemplate.getParametrizedText();
+                        int idx = text.indexOf('\n');
+                        label = convert(idx < 0 ? text : text.substring(0, idx) + DOTS, true);
+                    }
+                    String snippet = convert(codeTemplate.getParametrizedText(), false);
+                    codeActions.add(createCodeAction(Bundle.DN_SurroundWith() + label, CodeActionKind.RefactorRewrite, COMMAND_INSERT_SNIPPET, Collections.singletonMap(SNIPPET, snippet)));
+                }
+            }
+        }
+        return codeActions;
+    }
+
+    @Override
+    public Set<String> getCommands() {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        return CompletableFuture.completedFuture(false);
+    }
+
+    private static boolean isRightParenthesisOfLoopPresent(CompilationInfo info, int abbrevStartOffset) {
+        TokenHierarchy<?> tokenHierarchy = info.getTokenHierarchy();
+        TokenSequence<?> tokenSequence = tokenHierarchy.tokenSequence();
+        tokenSequence.move(abbrevStartOffset);
+        if (tokenSequence.moveNext()) {
+            TokenId tokenId = skipNextWhitespaces(tokenSequence);
```

Comment:
There are two `moveNext()` from the `abbrevStartOffset`, so the token at offset itself is skipped. OK ?

Reply:
From the `TokenSequence.move(int offset)` Javadoc: "Move token sequence to be positioned between `index-1`
and `index` tokens. .... Subsequent `moveNext()` or `movePrevious()` is needed to fetch a concrete token..."